Modules/_ctypes/_ctypes.c

```diff
@@ -967,18 +967,36 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
         return NULL;
     }
 #else
+	/* dlerror() always returns the latest error.
+	 *
+	 * Clear the previous value before calling dlsym(),
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+    dlerror();
     address = (void *)dlsym(handle, name);
-    if (!address) {
 #ifdef __CYGWIN__
-/* dlerror() isn't very helpful on cygwin */
+    if (!address) {
+        /* dlerror() isn't very helpful on cygwin */
         PyErr_Format(PyExc_ValueError,
                      "symbol '%s' not found",
                      name);
+        return NULL;
+    }
 #else
-        PyErr_SetString(PyExc_ValueError, dlerror());
-#endif
+    char *dlerr = dlerror();
+    if (dlerr) {
+		  // XXX: This assumes that UTF-8 is the default locale.
+		  //      Investigate if this can cause problems.
```

Comment:
Sorry, I might have screwed things up earlier with the indentation. (I made the review from my phone.)
There's some mixing of tabs and spaces here, so this should hopefully be the correct suggestion. My bad!
```suggestion
        // XXX: This assumes that UTF-8 is the default locale.
        // Investigate if this can cause problems.
```